modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+    if needsDeinit(t) then
+      _destroy(arg);
+  }
+
+  /*
+    Move-initialize a reference from a value. No assignment is performed
+    when moving ``rhs`` into the memory referred to by ``lhs``.
+
+    .. warning::
+
+      If the value at ``lhs`` is already initialized memory, it will be
+      overwritten by the contents of ``rhs`` without being deinitialized.
+      Call :proc:`explicitDeinit()` to deinitialize ``lhs`` first if
+      necessary.
+
+    :arg lhs: A reference to move-initialize
+    :type lhs: `?t`
+
+    :arg rhs: A 
+    :type rhs: `t`
+  */
+  proc moveInitialize(ref lhs: ?t, pragma "no auto destroy" in rhs: t) {
+    if lhs.type != nothing then
+      _move(lhs, rhs);
+  }
+
+  /*
+    Move the memory referred to by a reference into a new value.
+
+    .. warning::
+
+      The memory referred to by ``arg`` should be considered uninitialized
+      after a call to this function.
+
+    :arg arg: A reference to consume
+    :type arg: `?t`
+
+    :return: A reference moved into a new value
+    :rtype: `t`
+  */
+  proc moveToValue(const ref arg: ?t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var result: t;
+      _move(result, arg);
+      return result;
+    }
+  }
+
+  /*
+    Swap the values referred to by two references. No assignment takes
+    place when swapping ``lhs`` and ``rhs``.
+
+    :arg lhs: A value to swap
+    :type lhs: `?t`
+
+    :arg rhs: A value to swap
+    :type rhs: `t`
+  */
+  proc moveSwap(ref lhs: ?t, ref rhs: t) {
+    if t == nothing {
+      return none;
+    } else {
+      pragma "no init"
+      pragma "no copy"
+      pragma "no auto destroy"
+      var temp: t;
+      _move(temp, lhs);
+      _move(lhs, rhs);
+      _move(rhs, temp);
+    }
+  }
+
+  private inline proc _haltBadIndex(a, idx, indexName: string) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if !a.domain.contains(idx) then
+      boundsCheckHalt('Cannot move-initialize array because its domain ' +
+                      'does not contain: ' + indexName);
+  }
+
+  // Call to 'indexOrder' assumes 'idx' is a valid index.
+  private inline proc _haltBadElementRange(a, idx, numElements: int) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if numElements > a.size then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy exceeds array size');
+
+    if numElements <= 0 then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy is <= 0');
+
+    const order = a.domain.indexOrder(idx);
+    const hi = order + numElements;
+
+    if hi > a.size then
+      boundsCheckHalt('Cannot move-initialize array because one or ' +
+                      'more indices are outside its domain');
+  }
+
+  private inline proc _errorNonCopyableType(type t) {
+    if !isCopyableType(t) then
```

Comment:
Why wouldn't you be able to move initialize something that is not copyable?

Reply:
I guess because the alternative is saying that you're stealing elements from one array and putting them in another. Is that what this operation is doing? Just not how I visualized it.

Reply:
From reading all the other comments I'm assuming the answer is yes, it should always steal from `srcA`. So in that case should I just use a `_move` instead of `moveInitialize`?

Reply:
Sure